examples/ACE/utils/NNIAP.jl

```diff
@@ -78,6 +163,52 @@ function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w
     end
 end
 
+function learn!(nniap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w_e, w_f,_device)
+    if _device != gpu
+        return learn!(nniap, ds, opt, epochs, loss, w_e, w_f)
+    end
+    optim = Flux.setup(opt, nniap.nn)
+    nniap.nn = nniap.nn|> gpu
+    nniap.iap = nniap.iap|> gpu
+    ds  = ds |> gpu
+    optim = Flux.gpu(optim)
+
+    ∇gpu_loss(nn, iap, ds, w_e, w_f) = gradient((nn) -> gpu_loss(nn, iap, ds, w_e, w_f), nn)
+    losses = []
+    for epoch in 1:epochs
+        # Compute gradient with current parameters and update model
+        grads = ∇gpu_loss(nniap.nn, nniap.iap, ds, w_e, w_f)
+        Flux.update!(optim, nniap.nn, grads[1])
+        # Logging
+        curr_loss = gpu_loss(nniap.nn, nniap.iap, ds, w_e, w_f)
+        push!(losses, curr_loss)
+        println(curr_loss)
+    end
+end
+
+function learn!(nn, iap, ds, opt::Flux.Optimise.AbstractOptimiser, epochs, loss, w_e, w_f, true_energy, local_descriptors, _device=gpu, n_batches=1) # new
+    nniap = NNIAP(nn, iap)
+    optim = Flux.setup(opt, nniap.nn) |> _device
+    ∇loss(nn, iap, batch, true_energy, local_descriptors, _device, w_e, w_f) = gradient((nn) -> loss(nn, iap, batch, true_energy, local_descriptors, _device, w_e, w_f), nn)
+    losses = []
+    batch_lists = batch_and_shuffle(collect(1:length(ds)), n_batches)
+    batch_list_len = length(batch_lists)
+    for epoch in 1:epochs
+        true_energy = rand(Float32, 100) |> _device
+        local_descriptors = [rand(Float32, 26) for _ in 1:96]
```

Comment:
`local_descriptors`  need to be extracted from the current batch of atomic configurations. The local_descriptors type is Float32, so it is not necessary to cast them.